daml-lf/spec/daml-lf-1.rst

```diff
@@ -2569,6 +2575,150 @@ Map functions
 
   [*Available in versions >= 1.3*]
 
+
+Generic Map functions
+~~~~~~~~~~~~~~~~~~~~~
+
+* ``GENMAP_EMPTY : ∀ α. ∀ β. 'GenMap' α β``
+
+  Returns the empty generic map.
+
+  [*Available in versions >= 1.dev*]
+
+* ``GENMAP_INSERT : ∀ α. ∀ β.  α → β → 'GenMap' α β → 'GenMap' α β``
+
+  Inserts a new key and value in the map. If the key is already
+  present in the map, the associated value is replaced with the
+  supplied value.
+
+  This raises an error if the key is not a valid map key. Keys are
+  compared according to the rules listed below.
+
+  [*Available in versions >= 1.dev*]
+
+* ``GENMAP_LOOKUP : ∀ α. ∀ β.  α → 'GenMap' α β → 'Optional' α``
+
+  Lookups the value at a key in the map.
+
+  This raises an error if the key is not a valid map key. Keys are
+  compared according to the rules listed below.
+
+  [*Available in versions >= 1.dev*]
+
+* ``GENMAP_DELETE : ∀ α. ∀ β.  α → 'GenMap' α β → 'GenMap' α β``
+
+  Deletes a key and its value from the map. When the key is not a
+  member of the map, the original map is returned.
+
+  This raises an error if the key is not a valid map key. Keys are
+  compared according to the rules listed below.
+
+  [*Available in versions >= 1.dev*]
+
+* ``GENMAP_LIST : ∀ α. ∀ β.  'GenMap' α β → 'List' ⟨ key: α, value: β ⟩``
```

Comment:
@hurryabit Do we want to get ride of tuples ?

Reply:
If so, we could have separate `GENMAP_KEYS` and `GENMAP_VALUES` . All I did was copy the existing operations from `'Map'`

Reply:
IIRC, we had considered this for `TextMap` already but then decided against having `keys` and `values` separately. Unfortunately, I can't remember why. Doing the zipping in DAML land seems reasonable. If we go with the structural records, we would need to map them into proper DAML pairs anyway. Thus, the performance characteristics don't differ too much. Let's go with `GENMAP_KEYS` and `GENMAP_VALUES`.

Reply:
:+1: